File: ChatGPT/Simples/countingsort/countingsort.py
# How to implement the countingsort sort algorithm? in python

def counting_sort(arr):
    if not arr:
        return arr
    
    # Step 1: Find the range of input values
    max_val = max(arr)
    min_val = min(arr)
    range_of_elements = max_val - min_val + 1

    # Step 2: Initialize count array and output array
    count = [0] * range_of_elements
    output = [0] * len(arr)
    
    # Step 3: Count occurrences of each value
    for num in arr:
        count[num - min_val] += 1
    
    # Step 4: Modify count array to store cumulative counts
    for i in range(1, len(count)):
        count[i] += count[i - 1]
    
    # Step 5: Place elements in the output array
    for num in reversed(arr):
        output[count[num - min_val] - 1] = num
        count[num - min_val] -= 1
    
    return output

# Example usage
arr = [4, 2, 2, 8, 3, 3, 1]
sorted_arr = counting_sort(arr)
print(sorted_arr)  # Output: [1, 2, 2, 3, 3, 4, 8]


"""
Output:
    [1, 2, 2, 3, 3, 4, 8]
"""

File: ChatGPT/Simples/countingsort/CountingSort.java
// Input: How to implement the countingsort sort algorithm? in Java

import java.util.Arrays;

public class CountingSort {

    // Function to perform counting sort
    public static void countingSort(int[] array) {
        if (array == null || array.length == 0) {
            return;
        }

        // Step 1: Find the maximum value
        int max = Arrays.stream(array).max().getAsInt();

        // Step 2: Create the count array
        int[] count = new int[max + 1];

        // Step 3: Store the count of each number
        for (int num : array) {
            count[num]++;
        }

        // Step 4: Build the sorted output array
        int index = 0;
        for (int i = 0; i < count.length; i++) {
            while (count[i] > 0) {
                array[index++] = i;
                count[i]--;
            }
        }
    }

    // Main method to test the counting sort
    public static void main(String[] args) {
        int[] array = {4, 2, 2, 8, 3, 3, 1};
        System.out.println("Original array: " + Arrays.toString(array));
        countingSort(array);
        System.out.println("Sorted array: " + Arrays.toString(array));
    }
}

/*
Output:
    Original array: [4, 2, 2, 8, 3, 3, 1]
    Sorted array: [1, 2, 2, 3, 3, 4, 8]

Wrong:  is not stable
Score: 0.75
*/